src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form.ts

```diff
@@ -0,0 +1,189 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useState, useRef } from 'react';
+
+import { Form, FormData, FieldConfig, FieldsMap, FormConfig } from '../types';
+import { getAt, mapFormFields, unflattenObject, Subject } from '../lib';
+
+const DEFAULT_ERROR_DISPLAY_TIMEOUT = 500;
+
+export const useForm = <T = FormData>(
+  formConfig: FormConfig<T> | undefined = {}
+): { form: Form<T> } => {
+  const {
+    onSubmit,
+    schema,
+    defaultValue = {},
+    serializer = (data: any) => data,
+    deSerializer = (data: any) => data,
+    options = { errorDisplayDelay: DEFAULT_ERROR_DISPLAY_TIMEOUT, stripEmptyFields: true },
+  } = formConfig;
+  const defaultValueDeSerialized =
+    Object.keys(defaultValue).length === 0 ? defaultValue : deSerializer(defaultValue);
+  const [isSubmitted, setSubmitted] = useState(false);
+  const [isSubmitting, setSubmitting] = useState(false);
+  const [isValid, setIsValid] = useState(true);
+  const fieldsRefs = useRef<FieldsMap>({});
+
+  // formData$ is an observable we can subscribe to in order to receive live
+  // update of the raw form data. As an observable it does not trigger any React
+  // render().
+  // The <FormDataProvider> component is the one in charge of reading this observable
+  // and updating its state to trigger the necessary view render.
+  const formData$ = useRef<Subject<T>>(new Subject<T>({} as T));
+
+  // -- HELPERS
+  // ----------------------------------
+  const fieldsToArray = () => Object.values(fieldsRefs.current);
+
+  const stripEmptyFields = (fields: FieldsMap): FieldsMap => {
+    if (options.stripEmptyFields) {
+      return Object.entries(fields).reduce(
+        (acc, [key, field]) => {
+          if (field.value !== '') {
+            acc[key] = field;
+          }
+          return acc;
+        },
+        {} as FieldsMap
+      );
+    }
+    return fields;
+  };
+
+  // -- API
+  // ----------------------------------
+  const getFormData: Form<T>['__getFormData'] = (getDataOptions = { unflatten: true }) =>
+    getDataOptions.unflatten
+      ? (unflattenObject(
+          mapFormFields(stripEmptyFields(fieldsRefs.current), field => field.__getOutputValue())
+        ) as T)
+      : Object.entries(fieldsRefs.current).reduce(
+          (acc, [key, field]) => ({
+            ...acc,
+            [key]: field.__getOutputValue(),
+          }),
+          {} as T
+        );
+
+  const updateFormDataAt: Form<T>['__updateFormDataAt'] = (path, value) => {
+    const currentFormData = formData$.current.value;
+    formData$.current.next({ ...currentFormData, [path]: value });
+    return formData$.current.value;
+  };
+
+  const validateFields: Form<T>['__validateFields'] = async fieldNames => {
+    const fieldsToValidate = fieldNames
+      ? fieldNames.map(name => fieldsRefs.current[name]).filter(field => field !== undefined)
+      : fieldsToArray().filter(field => field.isPristine);
```

Comment:
Does this mean we're only validating fields which are pristine, AKA the user hasn't entered input into them? Would you mind explaining this? It seems like the opposite of what I would expect.

Reply:
Fields that are dirty have already been validated when their value changed. Thus their errors array contains already any possible error. So, to optimize, we only validate fields that haven't been touched by the user. I will add a comment.